File: src/algorithm_java/leetcode/Problem322.java
package algorithm.leetcode;

public class Problem322 {

    public static int coinChange(int[] coins, int amount) {
        int[] dp = new int[amount + 1];
        for (int i = 1 ; i <= amount; i++)
            dp[i] = amount + 1;
        dp[0] = 0;
        for (int i = 0; i <= amount; i++) {
            for (int j = 0; j < coins.length; j++) {
                if (i >= coins[j]) {
                    dp[i] = Math.min(dp[i], dp[i - coins[j]] + 1);
                }
            }
        }
        return dp[amount] > amount ? -1 : dp[amount];
    }

    public static void main(String[] args) {
        int[] coins = new int[]{1,2,5};
        System.out.println(coinChange(coins, 11));
    }

}

File: src/algorithm_cpp/problem322.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
	vector<int> dp(amount + 1, -1);
	dp[0] = 0;
	for (int i = 1; i <= amount; i++) {
	    for (int coin : coins) {
		if (coin <= i && dp[i - coin] != -1 && (dp[i - coin] + 1 < dp[i] || dp[i] == -1)) {
		    dp[i] = dp[i - coin] + 1;
		}
	    }
	}
	return dp[amount];
    }
};